### PR TITLE
Merging data type Map #15806

### PR DESCRIPTION
**Pull request by CurtizJ**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

The main difference from the original PR is that fixed duplicated sizes of arrays by using the type Nested.

### CODE REVIEW

File: tests/queries/0_stateless/01550_create_map_type.sql [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+insert into table_map select map('key1', number, 'key2', number * 2) from numbers(1111, 3);
+select a['key1'], a['key2'] from table_map;
+drop table if exists table_map;
+
+-- MergeTree Engine
+drop table if exists table_map;
+create table table_map (a Map(String, String), b String) engine = MergeTree() order by a;
+insert into table_map values ({'name':'zhangsan', 'gender':'male'}, 'name'), ({'name':'lisi', 'gender':'female'}, 'gender');
+select a[b] from table_map;
+select b from table_map where a = map('name','lisi', 'gender', 'female');
```
Comment (akuzm) [inline tests/queries/0_stateless/01550_create_map_type.sql:20]: Do we have a map constructor that accepts parallel arrays of keys/values a-la `sumMap`? And functions like `keys`/`values`/`items` that deconstruct it back into parallel arrays or an array of tuples. This constructor looks like it only works for a statically known set of keys, which might be not very practical.

Comment (CurtizJ) [reply tests/queries/0_stateless/01550_create_map_type.sql:20]: Map can be crated from tuple of arrays by using `CAST`. Like this: 
```sql
CAST(([1, 2, 3], ['1', '2', 'foo']), 'Map(UInt8, String)')
````. Maybe we need more convenient syntax to do that.

`keys` and `items` columns will be available after #17310 will be merged. `items` can be implemented also as a `CAST`.

File: src/Core/Field.h [hunk 2]
```diff
@@ -50,6 +50,7 @@ struct X : public FieldVector \
 
 DEFINE_FIELD_VECTOR(Array);
 DEFINE_FIELD_VECTOR(Tuple);
+DEFINE_FIELD_VECTOR(Map);
```
Comment (akuzm) [inline src/Core/Field.h:53]:
```suggestion
// An array of [key1, value1, key2, value2, ...]
DEFINE_FIELD_VECTOR(Map);
```

Comment (CurtizJ) [reply src/Core/Field.h:53]: But it is `[(key1, value1), (key2, value2), ...]`

Comment (akuzm) [reply src/Core/Field.h:53]: > But it is [(key1, value1), (key2, value2), ...]

Proves the need for a comment :laughing:

File: src/Columns/ColumnMap.h [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+#include <Core/Block.h>
+#include <Columns/ColumnArray.h>
+#include <Columns/ColumnVector.h>
+#include <Columns/ColumnTuple.h>
+
+namespace DB
+{
+
+/** Column, that is just group of 2 columns.
```
Comment (akuzm) [inline src/Columns/ColumnMap.h:11]: A more verbose description would help, e.g. it's a column of arrays of (key, value) tuples.

File: src/DataTypes/DataTypeMap.cpp [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+        throw;
+    }
+}
+
+void DataTypeMap::serializeText(const IColumn & column, size_t row_num, WriteBuffer & ostr, const FormatSettings & settings) const
+{
+    serializeTextImpl(column, row_num, ostr,
+        [&](const DataTypePtr & nested_type, const IColumn & nested_column, size_t pos)
+        {
+            nested_type->serializeAsTextQuoted(nested_column, pos, ostr, settings);
```
Comment (akuzm) [inline src/DataTypes/DataTypeMap.cpp:201]: Ahh so this `nested_type` is not the `nested` type that is `Array(Tuple(key,  value))`, but the key or value type that is passed to the writer. This confused me for a moment...

File: src/DataTypes/DataTypeMap.h [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  *
+  * Map's key and value only have types.
+  * If only one type is set, then key's type is "String" in default.
+  */
+class DataTypeMap final : public DataTypeWithSimpleSerialization
+{
+private:
+    DataTypePtr key_type;
+    DataTypePtr value_type;
+    DataTypePtr nested;
```
Comment (akuzm) [inline src/DataTypes/DataTypeMap.h:19]: A (redundant) comment would also be helpful here, e.g. the nested type is an `Array(Tuple(key, value))`.

File: src/DataTypes/FieldToDataType.cpp [hunk 6]
```diff
@@ -118,6 +119,24 @@ DataTypePtr FieldToDataType::operator() (const Tuple & tuple) const
...
+    DataTypes value_types;
+    key_types.reserve(map.size());
+    value_types.reserve(map.size());
+
+    for (const auto & elem : map)
+    {
+        const auto & tuple = elem.safeGet<const Tuple &>();
+        assert(tuple.size() == 2);
+        key_types.push_back(applyVisitor(FieldToDataType(), tuple[0]));
+        value_types.push_back(applyVisitor(FieldToDataType(), tuple[0]));
```
Comment (akuzm) [inline src/DataTypes/FieldToDataType.cpp:134]: Oops, should it be `tuple[1]`? Probably also needs some test coverage.

File: src/DataTypes/IDataType.h [hunk 7]
```diff
@@ -99,6 +101,8 @@ class IDataType : private boost::noncopyable
         /// Index of tuple element, starting at 1.
         String tuple_element_name;
 
+        String map_element_name;
```
Comment (akuzm) [inline src/DataTypes/IDataType.h:104]: Where is it set? Can't find on this GitHub page, maybe I'm missing some file...

**Review by akuzm — APPROVED** [review]
Left some minor comments above, overall looks good now.

**Comment by CurtizJ** [comment]
Fuzzer: #14974.

**Comment by CurtizJ** [comment]
Integration tests are broken in master.

**Comment by alexey-milovidov** [comment]
@CurtizJ One corner case has been found: #18492

File: src/Processors/Formats/Impl/ConstantExpressionTemplate.cpp [hunk 8]
```diff
@@ -196,6 +203,13 @@ class ReplaceLiteralsVisitor
                 if (not_null == array.end())
                     return true;
             }
+            else if (literal->value.getType() == Field::Types::Map)
+            {
+                const Map & map = literal->value.get<Map>();
+                if (map.size() % 2)
```
Comment (tavplubix) [inline src/Processors/Formats/Impl/ConstantExpressionTemplate.cpp:209]: What does this condition mean? Why is map size odd if there are NULL values?

Comment (CurtizJ) [reply src/Processors/Formats/Impl/ConstantExpressionTemplate.cpp:209]: But it not about null values. I think it's just an excessive check, which is actually unneeded.

File: src/DataTypes/DataTypeMap.cpp [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+
+            skipWhitespaceIfAny(istr);
+        }
+
+        offsets.push_back(offsets.back() + size);
+        assertChar('}', istr);
+    }
+    catch (...)
+    {
+        throw;
```
Comment (alexey-milovidov) [inline src/DataTypes/DataTypeMap.cpp:192]: WTF

**Comment by alexey-milovidov** [comment]
@akuzm a bug has been found: deserialization code is not exception safe.

**Comment by alexey-milovidov** [comment]
Fixed here: #36333

**Comment by alexey-milovidov** [comment]
@CurtizJ I don't understand why `ColumnMap` exists, maybe remove it?

**Comment by alexey-milovidov** [comment]
Does it exist only to enable Map type in Field?

**Pull request merged**